### PR TITLE
Report comments in MS Word with UIA enabled

### LINKED ISSUE #9285: Word 2016-365 installed with non-English language, with UIA is enabled, comments cannot be read
### Steps to reproduce:
Note, set office to use a language other than English.

approach 1:
1. Open the word document provided in the attachment.
2. When you arrow on a comment press insert+alt+c,

approach 2 - listing all comments in the elements list:
1. press insert+f7, when in the browse mode,
2. in the filter radio buttons category, find comments.

### Actual behavior:
When the first approach is used, the shortcut is not working, i.e. the comment was not read.
When you use the second approach, the following can be found in the log:

```
ERROR - queueHandler.flushQueue (19:54:49.137):
Error in func initElementType from eventQueue
Traceback (most recent call last):
  File "queueHandler.pyc", line 53, in flushQueue
  File "browseMode.pyc", line 902, in initElementType
  File "browseMode.pyc", line 922, in filter
  File "NVDAObjects\UIA\wordDocument.pyc", line 88, in label
TypeError: format() argument after ** must be a mapping, not NoneType
```

The comments are not listed

See code:
```
		if not obj.parent or obj.parent.name!='Comment':
			continue
```
The problem is that this object name is language sensitive i.e. in polish it is "Komentarz". So when the language in word interface is set to other than english, this condition is always true and NVDA doesn't see any comments even if there are some.

### Expected behavior:
The chomments should be listed and read, when nvda+alt+c, and when ins+f7 is pressed
#### NVDA installed/portable/running from source:
installed
#### NVDA version:
alpha-16768,a6f7fb40
#### Windows version:
windows 10, 1809, 17763
#### Name and version of other software in use when reproducing the issue:
microsoft office 365, 16.0.11231.20122
#### Other information about your system:
the sample files in the comments to follow
[Ęîëëîęâčóě_Ńňŕíĺ÷č÷.docx](https://github.com/nvaccess/nvda/files/2870461/Eieeieacoe_Nnril.c.docx)

#### Does the issue still occur after restarting your PC?
yes
#### Have you tried any other versions of NVDA?
no

### PR DESCRIPTION
**Pull request by jakubl7545**
### Link to issue number:
closes #9285 
### Summary of the issue:
With UIA enabled it is impossible to read comments. This is because comments ar filtered by its name which appeared to be different across languages.
Also when there are more than 13 comments dates can't be retrieved because don't fit on the screen.
### Description of how this pull request fixes the issue:
Instead of name comments are filtered by its role and AnnotationsPattern.
In situations when date is not on the screen only comment itself and author are reported.
### Testing performed:
Tested with NVDA running from source with word document which has comments like the attached one.
### Known issues with pull request:
None
### Change log entry:
Section: Bug fixes
It is now possible to read comments in MS Word with UIA enabled.
[test.docx](https://github.com/nvaccess/nvda/files/3227831/test.docx)

### CODE REVIEW

**Comment by zstanecic** [comment]
Hi,

Before merging this, can you provide the try build?

**Comment by lukaszgo1** [comment]
@zstanecic the build from this PR is here 
https://ci.appveyor.com/api/buildjobs/h0xaocd73ynxwnpa/artifacts/output%2Fnvda_snapshot_pr9631-17375%2C76080910.exe

File: source/NVDAObjects/UIA/wordDocument.py [hunk 1]
```diff
@@ -68,11 +70,16 @@ def getCommentInfoFromPosition(position):
 		UIAElement=UIAElementArray.getElement(index)
 		UIAElement=UIAElement.buildUpdatedCache(UIAHandler.handler.baseCacheRequest)
 		obj=UIA(UIAElement=UIAElement)
-		if not obj.parent or obj.parent.name!='Comment':
+		if (obj.parent.role !=controlTypes.ROLE_GROUPING and not
```
Comment (leonardder) [inline source/NVDAObjects/UIA/wordDocument.py]: Why did you remove the ``if not obj.parent` here? This breaks if the object has no parent.

File: source/NVDAObjects/UIA/wordDocument.py [hunk 2]
```diff
@@ -68,11 +70,16 @@ def getCommentInfoFromPosition(position):
 		UIAElement=UIAElementArray.getElement(index)
 		UIAElement=UIAElement.buildUpdatedCache(UIAHandler.handler.baseCacheRequest)
 		obj=UIA(UIAElement=UIAElement)
-		if not obj.parent or obj.parent.name!='Comment':
+		if (obj.parent.role !=controlTypes.ROLE_GROUPING and not
+		obj.parent.UIAElement.getCurrentPropertyValue(UIAHandler.UIA_IsAnnotationPatternAvailablePropertyId)):
 			continue
```
Comment (leonardder) [inline source/NVDAObjects/UIA/wordDocument.py]: I think it would be good to have some documentation about how you came up with the if clause above.

File: source/NVDAObjects/UIA/wordDocument.py [hunk 3]
```diff
@@ -68,11 +70,16 @@ def getCommentInfoFromPosition(position):
...
 		UIAElement=UIAElement.buildUpdatedCache(UIAHandler.handler.baseCacheRequest)
 		obj=UIA(UIAElement=UIAElement)
-		if not obj.parent or obj.parent.name!='Comment':
+		if (obj.parent.role !=controlTypes.ROLE_GROUPING and not
+		obj.parent.UIAElement.getCurrentPropertyValue(UIAHandler.UIA_IsAnnotationPatternAvailablePropertyId)):
 			continue
 		comment=obj.makeTextInfo(textInfos.POSITION_ALL).text
 		dateObj=obj.previous
 		date=dateObj.name
+		if not dateObj.previous:
```
Comment (leonardder) [inline source/NVDAObjects/UIA/wordDocument.py]: I think that it is more readable if you check obj.previous.previous first. Now, you are assigning an object to the dateObj variable and then, when it doesn't seem to be a dateObj, you're assigning the same object to authorObj.

For example:

```
		comment=obj.makeTextInfo(textInfos.POSITION_ALL).text
		tempObj = obj.previous.previous
		authorObj = tempObj else obj.previous
		author =authorObj.name
		if not tempObj:
			return dict(comment=comment,author=author)
		dateObj=obj.previous
		date=dateObj.name
		return dict(comment=comment,author=author,date=date)
```

File: source/NVDAObjects/UIA/wordDocument.py [hunk 4]
```diff
@@ -84,6 +91,9 @@ class CommentUIATextInfoQuickNavItem(TextAttribUIATextInfoQuickNavItem):
 	@property
 	def label(self):
 		commentInfo=getCommentInfoFromPosition(self.textInfo)
+		if len(commentInfo) ==2:
```
Comment (leonardder) [inline source/NVDAObjects/UIA/wordDocument.py]:
```suggestion
		if "date" not in commentInfo:
```

File: source/NVDAObjects/UIA/wordDocument.py [hunk 5]
```diff
@@ -320,6 +330,7 @@ class WordDocument(UIADocumentWithTableNavigation,WordDocumentNode,WordDocumentB
 	# Microsoft Word duplicates the full title of the document on this control, which is redundant as it appears in the title of the app itself.
 	name=u""
 
+	@script(gesture ="kb:NVDA+alt+c")
```
Comment (leonardder) [inline source/NVDAObjects/UIA/wordDocument.py]: This lacks the description. Please reuse the description for the script for the object model implementation of Word.

```suggestion
	@script(
		gesture="kb:NVDA+alt+c",
		# Translators: a description for a script that reports the comment at the caret.
		description=_("Reports the text of the comment where the System caret is located.")
	)
```

File: source/NVDAObjects/UIA/wordDocument.py [hunk 6]
```diff
@@ -334,17 +345,20 @@ def script_reportCurrentComment(self,gesture):
 			UIAElement=UIAElementArray.getElement(index)
```
Comment (leonardder) [inline source/NVDAObjects/UIA/wordDocument.py]: It really amazes me that this is not calling `getCommentInfoFromPosition`. Could you please change this in such a way that it uses that function and speaks a message based on what is in the dictionary? I think it is even possible to abstract this further and pas the result of CommentUIATextInfoQuickNavItem.label to ui.message.

If this sounds too complex, I"m happy to take this for you.

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
I haven't tested this myself, but here are some comments based on just reading the code.

File: source/NVDAObjects/UIA/wordDocument.py (hunk 1 above)
Comment (jakubl7545) [reply source/NVDAObjects/UIA/wordDocument.py]: I focused on replacing obj.name and missed it. Thanks for pointing this.

File: source/NVDAObjects/UIA/wordDocument.py (hunk 2 above)
Comment (jakubl7545) [reply source/NVDAObjects/UIA/wordDocument.py]: First I wanted to replace obj.name with something which doesn't change and decided for role which is grouping. In case it would be not enough I checked developer info for that object and discovered that comments have AnnotationPattern which may be significant for them. Then I found UIAElement.getCurrentPropertyValue checks for its presence. It requires one parameter which is propertyId. For AnnotationPattern it is 30118 but I didn't want to use that as itself it means nothing. Thats why as a parameter I used UIAHandler.UIA_IsAnnotationPatternAvailablePropertyId which provides that Id.
Thats how the process of creating looked like but I don't know how much of it should be documented. Maybe you have some ideas.

File: source/NVDAObjects/UIA/wordDocument.py (hunk 6 above)
Comment (jakubl7545) [reply source/NVDAObjects/UIA/wordDocument.py]: It works with getCommentInfoFromPosition but when I tried CommentUIATextInfoQuickNavItem.label I got its adress in a memory or "Property object is not callable" error.

File: source/NVDAObjects/UIA/wordDocument.py (hunk 2 above)
Comment (leonardder) [reply source/NVDAObjects/UIA/wordDocument.py]: I think it is ok as it is. Thanks for providing the details here. Leaving this treat unresolved until @michaeldcurran or someone else has looked at it.

File: source/NVDAObjects/UIA/wordDocument.py (hunk 6 above)
Comment (leonardder) [reply source/NVDAObjects/UIA/wordDocument.py]: I think it would be best to create a new _getPresentableCommentInfoFromPosition function that returns the label. Ten, you can make the label property return what that helper function returns for the particular textInfo, and same for the script.

Does that clear things up for you?

Comment (jakubl7545) [reply source/NVDAObjects/UIA/wordDocument.py]: I hope I understood you well. If so, should this new function has its doc string?
I also added a message for no comments in script.

**Comment by zstanecic** [comment]
For me this build works as expected.

Sorry for late reply.

**Comment by jakubl7545** [comment]
@leonardder Could you have another look on that?

**Comment by leonardder** [comment]
@jakubl7545: Could you merge current master into this?

**Comment by jakubl7545** [comment]
@leonardder: done.

**Comment by leonardder** [comment]
@jakubl7545 I'm sorry that it took me some time to come back to this. Could you plase merge master again and run `scons lint base=origin/master` to make sure this doesn't introduce linting errors?

**Comment by jakubl7545** [comment]
@leonardder It's done. Sorry for such a long delay.

**Review by LeonarddeR — DISMISSED** [review]
Thanks @jakubl7545 , looks good

**Comment by jakubl7545** [comment]
@michaelDCurran Would it be possible to merge this PR into 2019.3? It passed a review from @leonardder so I think it's ready.

**Comment by feerrenrut** [comment]
Thanks for resolving merge conflicts. Could you comment on any recent testing that you have done?

Specifically: 
- Comments in older versions of office
- Comments with UIA support disabled

**Comment by jakubl7545** [comment]
@feerrenrut Quick check shows that with NVDA 2019.3.1 and MS Word 2016 comments are read correctly when UIA support is disabled. However, I'm not sure whether such tests are necessary. Fixes in this PR are done in WordDocument.py which is used only when UIA support for word is enabled. As for older versions of Office I don't have one for testing but as far as I understand UIA support isn't available for older versions than 2016 anyway.

**Comment by dingpengyu** [comment]
hello
Will the current PR be merged into version 2020.2?
thank

File: source/NVDAObjects/UIA/wordDocument.py [hunk 7]
```diff
@@ -68,24 +70,37 @@ def getCommentInfoFromPosition(position):
 		UIAElement=UIAElementArray.getElement(index)
 		UIAElement=UIAElement.buildUpdatedCache(UIAHandler.handler.baseCacheRequest)
 		obj=UIA(UIAElement=UIAElement)
-		if not obj.parent or obj.parent.name!='Comment':
+		if not obj.parent or (
+			obj.parent.role != controlTypes.ROLE_GROUPING and not
```
Comment (feerrenrut) [inline source/NVDAObjects/UIA/wordDocument.py:74]: I thought Flake8 checked this, apparently not in this situation: please put the logical operators at the start of the line, it's easy to miss them when they are at the end.

File: source/NVDAObjects/UIA/wordDocument.py [hunk 8]
```diff
@@ -68,24 +70,37 @@ def getCommentInfoFromPosition(position):
 		UIAElement=UIAElementArray.getElement(index)
 		UIAElement=UIAElement.buildUpdatedCache(UIAHandler.handler.baseCacheRequest)
 		obj=UIA(UIAElement=UIAElement)
-		if not obj.parent or obj.parent.name!='Comment':
+		if not obj.parent or (
+			obj.parent.role != controlTypes.ROLE_GROUPING and not
+			obj.parent.UIAElement.getCurrentPropertyValue(UIAHandler.UIA_IsAnnotationPatternAvailablePropertyId)
```
Comment (feerrenrut) [inline source/NVDAObjects/UIA/wordDocument.py:75]: Please add a comment above each line answering "why?" 
It seems you have mostly answered this in response to @leonardder, putting it next to the code will likely save the next developer significant time.

However, from the comment it is not clear to me why the objects parent role must be group.

File: source/NVDAObjects/UIA/wordDocument.py [hunk 9]
```diff
@@ -327,31 +342,17 @@ def event_UIA_notification(self, activityId=None, **kwargs):
...
-			# Translators: The message reported for a comment in Microsoft Word
-			ui.message(_("{comment} by {author} on {date}").format(comment=comment,date=date,author=author))
-			return
-
-	__gestures={
-		"kb:NVDA+alt+c":"reportCurrentComment",
-	}
+		commentInfo = getCommentInfoFromPosition(caretInfo)
+		if commentInfo is not None:
+			ui.message(getPresentableCommentInfoFromPosition(commentInfo))
```
Comment (feerrenrut) [inline source/NVDAObjects/UIA/wordDocument.py:354]: Great code re-use!

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/NVDAObjects/UIA/wordDocument.py (hunk 7 above)
Comment (jakubl7545) [reply source/NVDAObjects/UIA/wordDocument.py:74]: Should I move "or (" from line 73 or only "and not" from line 74?

File: source/NVDAObjects/UIA/wordDocument.py (hunk 8 above)
Comment (jakubl7545) [reply source/NVDAObjects/UIA/wordDocument.py:75]: @feerrenrut It checks for grouping because from the log viewer I see that the role of that object is grouping.
Are such comments ok?
			# Because the name of this object is language sensitive check for its role instead.
obj.parent.role != controlTypes.ROLE_GROUPING
			# And to make sure it's the right object check if it has Annotation Pattern
and not obj.parent.UIAElement.getCurrentPropertyValue(UIAHandler.UIA_IsAnnotationPatternAvailablePropertyId)

File: source/NVDAObjects/UIA/wordDocument.py (hunk 7 above)
Comment (feerrenrut) [reply source/NVDAObjects/UIA/wordDocument.py:74]: Just the "and not"

File: source/NVDAObjects/UIA/wordDocument.py (hunk 8 above)
Comment (feerrenrut) [reply source/NVDAObjects/UIA/wordDocument.py:75]: Do you know for sure that all comment objects have a parent with role grouping? Could you theoretically remove this check and just check for `not obj.parent.UIAElement.getCurrentPropertyValue(UIAHandler.UIA_IsAnnotationPatternAvailablePropertyId` if not, why not?

It is also acceptable to keep the role_grouping check for performance reasons. For instance, if we knew that when there is no role_grouping then there will definitely not be `obj.parent.UIAElement.getCurrentPropertyValue(UIAHandler.UIA_IsAnnotationPatternAvailablePropertyId` then we can use `role_grouping` as a shortcut to avoid a UIA call.

Comment (jakubl7545) [reply source/NVDAObjects/UIA/wordDocument.py:75]: I removed role check and it still works. I doubt that role is connected with UIA pattern so maybe the best way here is to check only for annotation pattern.

Comment (feerrenrut) [reply source/NVDAObjects/UIA/wordDocument.py]: I'd like to get advice from @michaelDCurran here. Avoiding UIA calls where possible helps performance, but I don't want that to come at a cost to accuracy.

Comment (jakubl7545) [reply source/NVDAObjects/UIA/wordDocument.py]: For now I deleted role check.

**Comment by codeofdusk** [comment]
Any updates on this PR? CC @michaelDCurran.

**Comment by lukaszgo1** [comment]
This has been reported several times and it looks like this PR is just awaiting review. @michaelDCurran Do you intent to look at this one?

**Comment by michaelDCurran** [comment]
I'm quite worried that this implementation might have broken with newer 
builds of Office 365. I shall give it a test and report back.

**Comment by cary-rowen** [comment]
What is the latest news of this PR and is it possible to be merged in 2020.4?
Thanks.

**Comment by lukaszgo1** [comment]
@michaelDCurran Ping. This has been asked about again in #11989

**Comment by cary-rowen** [comment]
@michaelDCurran Can you review this

**Comment by michaelDCurran** [comment]
This is not reporting comments for me in MS Word build 16.0.13707 nor has it been for many months. Can someone else confirm that this code still works in Microsoft Word for them? We need to also work out what MS Word version this broke in, and perhaps try and come up with a solution that works for both.

**Comment by michaelDCurran** [comment]
This pr does not work for my with Microsoft Word build 16.0.13707, or has it worked with builds of MS Word that I have had for some months now. I am right now looking into how we can make this code more robust. At very least, as the annotation pattern is available, we can actually use that pattern to get the needed info directly now.

**Comment by michaelDCurran** [comment]
Here is an updated copy of getCommentInfoFromPosition which uses annotation properties directly, and does not need to walk as many options. This code works on my 16.0.13707 build of MS Word:
```
def getCommentInfoFromPosition(position):
	"""
	Fetches information about the comment located at the given position in a word document.
	@param position: a TextInfo representing the span of the comment in the word document.
	@type L{TextInfo}
	@return: A dictionary containing keys of comment, author and date
	@rtype: dict
	"""
	val=position._rangeObj.getAttributeValue(UIAHandler.UIA_AnnotationObjectsAttributeId)
	if not val:
		return
	try:
		UIAElementArray=val.QueryInterface(UIAHandler.IUIAutomationElementArray)
	except COMError:
		return
	for index in range(UIAElementArray.length):
		UIAElement=UIAElementArray.getElement(index)
		UIAElement=UIAElement.buildUpdatedCache(UIAHandler.handler.baseCacheRequest)
		typeID = UIAElement.GetCurrentPropertyValue(UIAHandler.UIA_AnnotationAnnotationTypeIdPropertyId)
		if typeID == UIAHandler.AnnotationType_Comment:
			comment = UIAElement.GetCurrentPropertyValue(UIAHandler.UIA_NamePropertyId)
			author = UIAElement.GetCurrentPropertyValue(UIAHandler.UIA_AnnotationAuthorPropertyId)
			date = UIAElement.GetCurrentPropertyValue(UIAHandler.UIA_AnnotationDateTimePropertyId)
			return dict(comment=comment,author=author,date=date)
```

However, as MS Word now supports comment threads, much of the time, the comment reported will just be "comment thread started by Michael Curran" or similar. Users should really now start viewing comments in the comment thread pain.

**Comment by jakubl7545** [comment]
> 
> Can someone else confirm that this code still works in Microsoft Word for them

@michaelDCurran Yes, it works for me in MSO (16.0.13530.20368). It reads comment itself, author and date exactly as it should.

**Comment by cary-rowen** [comment]
It works fine for me: Microsoft 365 MSO (16.0.13530.20132) 64 Bit
@michaelDCurran

**Comment by jakubl7545** [comment]
@michaelDCurran I've tried your code but it didn't work. Probably because AnnotationType_Comment is not available. Only AnnotationType_Unknown is returned. So I've combined both approaches to ensure that                 comments are read in different versions of Office. Could you check if it works for you now?

**Review by michaelDCurran — APPROVED** [review]

**Comment by michaelDCurran** [comment]
@jakubl7545  Yes, this is working for me. Thanks for your work on this.

**Comment by michaelDCurran** [comment]
@jakubl7545  I think there may be one or more review actions from @feerrenrut  before he can also approve?

**Review by feerrenrut — APPROVED** [review]
All changes addressed

**Comment by cary-rowen** [comment]
In version Microsoft 365 MSO (16.0.14228.20200) 64-bit  the pr reported an incorrect comment:
When I moved the cursor to the position containing the comment and pressed NVDA + Alt + C, NVDA reported the following:
"Comment: Resolved Comment Hint Button by Rowen Cary on 25 minutes ago"
According to the above, the content of the comment is: "Resolved Comment Hint" This is wrong and not the actual comment I wrote.
When I turned off the UIA mode in Word, everything returned to normal.
My language is Simplified Chinese, but I installed the English version of Office and can also reproduce this problem.

@feerrenrut 
@michaelDCurran

**Comment by lukaszgo1** [comment]
@cary-rowen  Please create a new issue for this and preferably attach a sample document reproducing the problem.
